File: Rcode/ARIMA1.R
#ライブラリインストール(必要に応じて)
#install.packages( "tidyverse" )
#install.packages( "fable" )
#install.packages( "feasts" )
#install.packages( "urca" )
#install.packages( "reshape2" )

# 該当リポジトリを変数に格納
repo = 
  c( "https://raw.githubusercontent.com/u-10bei/Population_Projections/" )

# 人口推計に使うデータの格納場所を変数に格納
popURL = 
  c( "main/data/population_jp_year.csv" )

# ライブラリの読み込み
library( readr )

repo |>
  paste0( popURL ) |>                     # 読み込むアドレスの編集
  read_csv( show_col_types = FALSE ) ->   # ネット上のファイル読み込み
pop_df

# ライブラリの読み込み
library( ggplot2 )

# 総人口のグラフ
pop_df |>
  ggplot( aes( x = Year,
               y = Total )) +
  geom_line()

# 自己相関のグラフ
pop_df$Total |>
  acf()

# 偏自己相関のグラフ
pop_df$Total |>
  pacf() 

# 学習データと予測データ
prow_test = 5

pop_df |>
  tail( n = prow_test ) ->
pop_test

# ライブラリの読み込み
library( fable )

pop_df |>
  head( n = { nrow( pop_df ) - prow_test }) |>
  as_tsibble( index = Year ) -> 
pop_train

# ライブラリの読み込み
library( dplyr )

# ＡＩＣが自動推定＋２までのモデルを候補として列挙
capture.output({
  pop_train |>
    model(
      arima = ARIMA( Total,
                     ic = "aic",
                     trace = TRUE,
                     stepwise = FALSE ))
  }) |>
  read_table( col_names = c( "symbol", "AIC" ),
              col_types = "cn",
              comment = "<" ) |>
  na.omit() ->
trace_arima

max_AIC = min( trace_arima$AIC ) + 2
trace_arima |>
  filter( AIC < max_AIC ) 

# 予測
pop_train |>
  model(
    arima120 = ARIMA( Total ~ 0 + pdq( 1, 2, 0 )),
    arima220 = ARIMA( Total ~ 0 + pdq( 2, 2, 0 )), 
    arima021 = ARIMA( Total ~ 0 + pdq( 0, 2, 1 )),
    arima121 = ARIMA( Total ~ 0 + pdq( 1, 2, 1 )),
    arima221 = ARIMA( Total ~ 0 + pdq( 2, 2, 1 )),
    arima022 = ARIMA( Total ~ 0 + pdq( 0, 2, 2 ))) |>
  forecast( h = "5 years" ) ->
pop_arima_f
pop_arima_f |>
  filter( Year == 2020 ) 
pop_test |>
  select( Year, Total ) |>
  tail( 1 )

# 社人研予測との比較
# 該当ＵＲＬを変数に格納
ipssURL = 
  c( "main/data/forecast_ipss.csv" )

repo |>
  paste0( ipssURL ) |>                    # 読み込むアドレスの編集
  read_csv( show_col_types = FALSE ) ->   # ネット上のファイル読み込み
ipss_test

pop_arima_f |>
  as.data.frame() |>
  filter( .model == "arima221" ) |>
  select( Year, "forecast" = .mean ) |>
  inner_join( pop_test, by = "Year" ) |>
  inner_join( ipss_test, by = "Year" )|>
  select( Year,
          Total,
          forecast,
          DMBM,
          DMBH,
          DLBM,
          DLBH ) ->
join_test
join_test

# ライブラリの読み込み
library( reshape2 )

# 描画
join_test |> 
  melt(id = "Year",
       measure = c( "Total",
                    "forecast",
                    "DMBM",
                    "DMBH",
                    "DLBM",
                    "DLBH")) |>
  ggplot( aes( x = Year,
               y = value,
               shape = variable,
               colour = variable,
               group = variable )) +
  geom_line() +
  geom_point()

File: Pythoncode/ARIMA1.py
# 該当ＵＲＬを変数に格納
repo = "https://raw.githubusercontent.com/u-10bei/Population_Projections/"
popURL = repo + "main/data/population_jp_year.csv"

# ライブラリの読み込み
import pandas as pd
from dateutil.relativedelta import relativedelta

# ネット上のファイル読み込み
pop_csv = pd.read_csv( popURL,
                      parse_dates = [ 'Year' ])
pop_csv.Year = pop_csv.Year.apply( lambda x: x + relativedelta( months = 9 ))
pop_df = pop_csv[[ 'Year', 'Total' ]]\
          .astype({ 'Total': float }).set_index( 'Year' )

# ライブラリの読み込み
import matplotlib.pyplot as plt
plt.style.use( 'ggplot' )

# グラフのサイズ変更
plt.rcParams[ 'figure.figsize' ] = [ 10, 5 ]

# 総人口のグラフ
plt.plot( pop_df )

# ライブラリの読み込み
import pmdarima as pm

# 自己相関のグラフ
pm.utils.plot_acf( pop_df,
                   lags = 20,
                   alpha = .05 )

# 偏自己相関のグラフ
pm.utils.plot_pacf( pop_df,
                    lags = 20,
                    alpha = .05 )

# 学習データと予測データ
prow_test = 5
pop_test = pop_df.tail()
pop_train = pop_df.head( prow_test * -1 )

# ＡＲＩＭＡモデルの推定
pop_arima = pm.auto_arima( pop_train,
                           information_criterion = 'aic' )
pop_arima

# モデリング
from statsmodels.tsa.arima.model import ARIMA
pop_arima = ARIMA( pop_train, order = ( 0, 2, 0 )).fit()

# ＡＲＩＭＡによる予測
pop_arima_f = pop_arima.predict( '2016', '2020' )
pop_arima_f

# 社人研予測との比較
# 該当ＵＲＬを変数に格納
ipssURL = repo + "main/data/forecast_ipss.csv"

# ネット上のファイル読み込み
ipss_csv = pd.read_csv( ipssURL,
                        parse_dates = [ 'Year' ])
ipss_csv.Year = ipss_csv.Year.apply( lambda x: x + relativedelta( months = 9 ))
ipss_test = ipss_csv.set_index( 'Year' )

# ライブラリの読み込み
import numpy as np

# 予測と実測の比較（グラフ）
x_axis = np.arange(pop_arima_f.shape[0])
plt.plot( x_axis,
          pop_test,
          label = "Total",
          color = 'black' )
plt.plot( x_axis,
          pop_arima_f,
          label="forecast",
          color='blue')
plt.plot( x_axis,
          ipss_test.DMBM,
          label="DMBM",
          color='green')
plt.plot( x_axis,
          ipss_test.DMBH,
          label="DMBH",
          color='green')
plt.plot( x_axis,
          ipss_test.DLBH,
          label="DLBH",
          color='green')
plt.plot( x_axis,
          ipss_test.DLBM,
          label="DLBM",
          color='red')
plt.legend()
plt.show()